namer/consul/src/main/scala/io/buoyant/namer/consul/DcServices.scala

```diff
@@ -0,0 +1,96 @@
+package io.buoyant.namer.consul
+
+import com.twitter.finagle.http.{Response, Status}
+import com.twitter.finagle.{Addr, Failure}
+import com.twitter.util._
+import io.buoyant.consul.v1
+
+private[consul] object DcServices {
+
+  /**
+   * Contains all cached serviceMap responses and the mapping of names
+   * to Addrs for a particular datacenter.
+   */
+  def apply(
+    consulApi: v1.ConsulApi,
+    name: String,
+    domain: Option[String]
+  ): Activity[Map[SvcKey, Var[Addr]]] = {
+
+    def getServices(index: Option[String]): Future[v1.Indexed[Set[SvcKey]]] =
+      consulApi.serviceMap(
+        datacenter = Some(name),
+        blockingIndex = index,
+        retry = true
+      ).map(toServices)
+
+    val states = Var.async[Activity.State[Map[SvcKey, Var[Addr]]]](Activity.Pending) { state =>
+      @volatile var stopped: Boolean = false
+
+      def loop(index0: Option[String], cache: Map[SvcKey, Var[Addr]]): Future[Unit] =
+        if (stopped) Future.Unit
+        else getServices(index0).transform {
+          case Throw(e@v1.UnexpectedResponse(rsp)) if rsp.status == Status.NotFound =>
```

Comment:
Yeah, unfortunately it doesn't look like there is a way to distinguish between non-existing datacenter and temporarily disconnected one. It's not clear whether `/v1/catalog/datacenters` lists only datacenters availbale right now or any "ever seen". I think it worth an investigation so maybe put a TODO for now?..